### PR TITLE
dev-require_resources

### PR DESCRIPTION
**Pull request by bio-boris**
Creating a PR to test codacy

I suspect it may contain some cruft.

# Description of PR purpose/changes

* Update readme
* Add documentation to ee2 spec
* Refactor auth tests to be unit testable

# Jira Ticket / Github Issue #
- [ ] Added the Jira Ticket to the title of the PR e.g. (DATAUP-69 Adds a PR template)

# Testing Instructions
* Details for how to test the PR: 
- [ ] Tests pass in travis and locally 
- [ ] Changes available by spinning up a local test suite and doing X

# Dev Checklist:

- [ ] My code follows the guidelines at https://sites.google.com/truss.works/kbasetruss/development
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have run Black and Flake8 on changed Python Code manually or with git precommit (and the travis build passes)

# Updating Version and Release Notes (if applicable)

- [ ] [Version has been bumped](https://semver.org/) for each release
- [ ] [Release notes](/RELEASE_NOTES.md) have been updated for each release (and during the merge of feature branches)

### CODE REVIEW

File: execution_engine2.spec [hunk 1]
```diff
@@ -525,30 +525,59 @@
...
+            projection - the list of fields included in the returned job. By default all fields.
+            limit - the maximum number of jobs returned.
+            sort_order - the order in which the results were sorted by the job ID - + for
+                ascending, - for descending.
+
+            TODO: DOCUMENT THE RETURN OF STATS mapping
         */
         typedef structure {
-            mapping<job_id, JobState> jobs;
+            list<JobState> jobs;
```
Comment (bio-boris) [inline execution_engine2.spec:544]: verified, correct

File: execution_engine2.spec [hunk 2]
```diff
@@ -525,30 +525,59 @@
...
             int count;
             int query_count;
-            list<string> filter;
+            mapping<string, string> filter;
```
Comment (bio-boris) [inline execution_engine2.spec:547]: verified, correct

File: execution_engine2.spec [hunk 3]
```diff
@@ -525,30 +525,59 @@
...
-
         /*
           Check job for all jobs in a given date/time range for all users (Admin function)
-            float start_time; # Filter based on creation timestamp since epoch
-            float end_time; # Filter based on creation timestamp since epoch
-            list<string> projection; # A list of fields to include in the projection, default ALL See "Projection Fields"
-            list<string> filter; # A list of simple filters to "AND" together, such as error_code=1, wsid=1234, terminated_code = 1
-            int limit; # The maximum number of records to return
-            string user; # Optional. Defaults off of your token
+            Notes on start_time and end_time:
```
Comment (bio-boris) [inline execution_engine2.spec:556]: Looks good

File: lib/execution_engine2/authorization/authstrategy.py [hunk 4]
```diff
@@ -10,76 +10,67 @@
...
+    :param level: string - if "read", then returns the read value, if "write", return whether
+         the user can write.
     :returns: bool - True if the permission is valid, False otherwise.
     """
     if user_id == job.user:
         return True
     if job.authstrat == KBASE_WS_AUTHSTRAT:
         if job.wsid is None:
             return False
-        ws_auth = WorkspaceAuth(token, user_id, config["workspace-url"])
```
Comment (bio-boris) [inline lib/execution_engine2/authorization/authstrategy.py:82]: Looks good

File: lib/execution_engine2/authorization/authstrategy.py [hunk 5]
```diff
@@ -89,15 +80,15 @@ def _check_permissions(
...
+    jobs: List[Job], user_id: str, ws_auth: WorkspaceAuth, level="read"
 ) -> List[bool]:
     """
     Returns True for each job the user has read access to, and False for the ones they don't.
     :param job: a Job model object
     :param user_id: string - the user id
-    :param token: string - the user's auth token
-    :param config: dict - the service config
-    :param level: string - if "read" then tests if the Job can be read, otherwise checks if it can be written
+    :param ws_auth: a workspace authorization instance initialized with the user's token.
```
Comment (bio-boris) [inline lib/execution_engine2/authorization/authstrategy.py:89]: probably don't need a period

File: test/tests_for_auth/ee2_authstrategy_test.py [hunk 6]
```diff
@@ -73,13 +76,18 @@ def _mock_ws_deleted(self, rq_mock, ws_id):
             "POST", self.ws_url, [{"json": response, "status_code": 500}]
         )
 
+    def _get_workspace_auth(self, token) -> WorkspaceAuth:
+        # TODO these tests can be converted to unit tests by mocking the WorkspaceAuth class
```
Comment (bio-boris) [inline test/tests_for_auth/ee2_authstrategy_test.py:80]: Looks good

File: lib/execution_engine2/authorization/workspaceauth.py [hunk 7]
```diff
@@ -1,8 +1,8 @@
 from typing import List, Dict
 from enum import Enum
-from lib.execution_engine2.authorization.basestrategy import AuthStrategy
```
[the remark this answers is not in the document]
Comment (MrCreosote) [reply lib/execution_engine2/authorization/workspaceauth.py:3]: Good to hear, I was wondering why it was there. It was definitely causing problems with the mock libraries on my end, as the `@patch` method seemed to be patching the wrong thing

**Comment by MrCreosote** [comment]
Hmm, codeclimate is failing but none of the failures are on code I changed

**Pull request merged**